### PR TITLE
[Merged by Bors] - feat: string diagram widget

### PR DESCRIPTION
**Pull request by yuma-mizuno**
Add the widget to display string diagram of monoidal category morphism.

---

### Issues

- ~~The widget fails to make a diagram for `f ▷ (X ⊗ Y)` without errors, although it succeeds for `(α_ X X Y).inv ≫ f ▷ X ▷ Y ≫ (α_ Y X Y).hom`. I expect the former expression to be rewritten to the latter by `eval`. There may be a bug somewhere, but I can't find it.~~ The problem was that the widget program can't find implicit arguments when using `mkAppM` (e.g. it can't find a `Category C` instance). I don't understand why this happens, but I removed all `mkAppM` and then the program works.

### Future plans (in another PR)

- [x] For `tensorHom`, the widget simply expand it by `tensorHom_def`. A better way is to allow `tensorHom` between non-composite morphisms, but this requires more works.  See #11080.
- [x] Supporting 2-morphisms in bicategories. See #12107.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: widget/src/penrose/monoidal.sty
Comment (eric-wieser) [inline widget/src/penrose/monoidal.sty:48]: I could barely tell the difference between these; changing cornerRadius or remove the background color might make them easier to differentiate, assuming that's actually desirable.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/StringDiagram.lean:63]: My inclination would be to rewrite this with Qq to make it more robust to metavariables / abbreviations / metadata.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/StringDiagram.lean:286]: I had a go at Qq-ifying this, but the best I could come up with was
```
/-- Evaluate the expression of a 2-morphism into a normalized form. -/
partial def eval {u v : Level} {α : Q(Type u)} {C : Q(Category.{v} $α)} {M : Q(MonoidalCategory $α)}
    {x y : Q($α)} (e : Q($x ⟶ $y)) : MetaM NormalExpr := do
  if let .defEq (h : $x =Q $y) ← isDefEqQ x y then
    if let ~q(𝟙 _) := e then
      return NormalExpr.id (toMor₁ x)
  if let ~q($f ≫ $g) := e then
    return evalComp (← eval f) (← eval g)
  if let ~q($xl ⊗ $xr) := x then
    if let ~q($yl ⊗ $yr) := y then
      if let .defEq (h : $xl =Q $yl) ← isDefEqQ xl yl then
        if let ~q(_ ◁ $η) := e then
          evalWhiskerLeftExpr (toMor₁ xl) (← eval η)
      if let .defEq (h : $xr =Q $yr) ← isDefEqQ xr yr then
        if let ~q($η ▷ _) := e then
          evalWhiskerRightExpr (← eval η) (toMor₁ xr)
  return NormalExpr.of e
  ```
  which doesn't actually work thanks to https://github.com/leanprover-community/quote4/issues/21. So I think Qq-ification is probably not on the cards after all.

Comment (yuma-mizuno) [reply Mathlib/Tactic/Widget/StringDiagram.lean:286]: Thank you very much for trying it! I knew very little about how to use Qq.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/StringDiagram.lean:332]: I think this is `List.enumFrom`?

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/StringDiagram.lean:336]: `List.enum`

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (yuma-mizuno) [reply Mathlib/Tactic/Widget/StringDiagram.lean]: Oh, I missed it, thanks!

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/StringDiagram.lean:280]: Can you add some tests for this in the test file? I think the tests will be easiest with Qq, as something like:
```
variable {u v : Level} {C : Q(Type u)} {iC : Q(Category.{v} $C)} {iM : Q(MonoidalCategory $C)}
variable {X Y : Q($C)} (η : Q(𝟙_ _ ⟶ $X ⊗ $Y)) (ε : Q($Y ⊗ $X ⟶ 𝟙_ _))

#guard_msgs
#eval eval q(η ▷ X ≫ (α_ _ _ _).hom ≫ X ◁ ε)
```

File: test/StringDiagram.lean
Comment (eric-wieser) [inline test/StringDiagram.lean:49]: This is nicer than my Qq suggestion anyway, thanks :)

File: test/StringDiagram.lean
Comment (eric-wieser) [inline test/StringDiagram.lean:43]: convention is to use
```suggestion
elab "normalize% " t:term : term => do
```
to make it clear that `normalize` is magic and not a regular function

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/StringDiagram.lean:22]: Can you add a docstring to each of these constructors?

**Comment by eric-wieser** [comment]
In this example:

image

Should the middle line be labelled `Y`?

**Comment by yuma-mizuno** [comment]
> In this example:
> 
> image
> 
> Should the middle `\` line be labelled `Y`?

Yes. Currently, the widget does not put labels to strings, but I think it is possible to add such a feature by editting the `monoidal.sty` file.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/StringDiagram.lean:34]:
```suggestion
inductive Mor₁ : Type
  /-- `id C` is the expression for `𝟙_ C`. -/
  | id : Expr → Mor₁
  /-- `comp X Y` is the expression for `X ⊗ Y` -/
  | comp : Mor₁ → Mor₁ → Mor₁
```

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/StringDiagram.lean:73]: Can you give the Lean expression for each case in the comment, like you did above?

**Comment by yuma-mizuno** [comment]
Now the widget supports `monoidalComp`. To support this, I redefined the normal form of 2-morphisms. The new normal form has the form `α₀ ≫ η₀ ≫ α₁ ≫ η₁ ≫ ... αₙ ≫ ηₙ ≫ αₙ₊₁`, where `αᵢ` are structural 2-morphisms and `ηᵢ` are non-structural 2-morphisms. See `NormalExpr`.

**Comment by eric-wieser** [comment]
It seems a lot of this PR is about normalization in monoidal categories; is there a `moncat_normalize` tactic hiding here? Maybe such a tactic is out of scope for this Pr, but it would be good to put the content about normalization and the content about rendering in separate files, if such a split seems reasonable to you.

**Comment by yuma-mizuno** [comment]
Your suggestion makes sense to me. I will make a separate file that deals with normalization.

**Comment by yuma-mizuno** [comment]
cf. #11133

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (Vtec234) [inline Mathlib/Tactic/Widget/StringDiagram.lean:312]: Instead of having two presenters, how about having just one that shows both sides? That way it's easy to see the difference and what needs to be done to match the two sides up. You could do it like this:
```lean
@[expr_presenter]
def stringPresenter : ExprPresenter where
  userName := "Equality of string diagrams"
  layoutKind := .block
  present type := do
    let some d ← stringLeftM? type
      | throwError "Couldn't find a string diagram."
    let some e ← stringRightM? type
      | throwError "Couldn't find a string diagram."
    return <div className="flex">
      <div className="w-50">{d}</div>
      <div className="w-50">{e}</div>
    </div>
```

File: test/StringDiagram.lean
Comment (Vtec234) [inline test/StringDiagram.lean:20]: It's okay to have tests that use it, but afaict it's ultimately a bit awkward to use `SelectionPanel` to show the goal type because it requires the extra click. `GoalTypePanel` is preferrable.

File: test/StringDiagram.lean
Comment (Vtec234) [inline test/StringDiagram.lean:26]: Instead of writing `with_panel_widgets` everywhere, you can also use a single `show_panel_widgets [local GoalTypePanel]` command towards the top of the file.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (Vtec234) [inline Mathlib/Tactic/Widget/StringDiagram.lean:193]:
```suggestion
```
As it's not used.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (Vtec234) [inline Mathlib/Tactic/Widget/StringDiagram.lean:257]: I wonder, since `DiagramBuilderM` is now being used in more places, maybe it should be upstreamed to ProofWidgets?

Comment (yuma-mizuno) [reply Mathlib/Tactic/Widget/StringDiagram.lean:254]: That sounds reasonable. Should I make such a PR?

Comment (Vtec234) [reply Mathlib/Tactic/Widget/StringDiagram.lean:254]: I already put it there :) However, this PR won't be able to use it until mathlib4 uses the next release of ProofWidgets4.

**Comment by semorrison** [comment]
@yuma-mizuno, would you mind going through and making conversations above as resolved where appropriate, to make it easier for someone new to catch up on the current state?

**Comment by eric-wieser** [comment]
I think it would make sense to split out the `MonoidalCategory` normalization stuff to it's own PR, on which this one can depend?

**Comment by yuma-mizuno** [comment]
@eric-wieser It's probably #11133. This should be merged into master soon, so I merged it here in advance.

**Comment by yuma-mizuno** [comment]
It seems that we need to deal with a new monad hierarchy.

**Comment by eric-wieser** [comment]
Ah, you didn't mark it as a dependency of this PR (see the comment in the PR template) so I didn't notice that one!

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (yuma-mizuno) [reply Mathlib/Tactic/Widget/StringDiagram.lean:274]: @Vtec234 Now I need a slightly generalized version of `DiagramBuilderM` (see https://github.com/leanprover-community/mathlib4/pull/10581/commits/035db686c6e5cb737315da22f636304e39e26ee0). Would you mind if I make a PR to ProofWidget4 about this change?

**Comment by yuma-mizuno** [comment]
I've resolved the conversation so far. The latest commit contains some new code that was needed after merging #11133.

File: test/StringDiagram.lean
Comment (joneugster) [inline test/StringDiagram.lean:21]: Can you leave a comment mentioning why the RHS has no string diagram

File: Mathlib/Tactic/CategoryTheory/Monoidal.lean
Comment (joneugster) [inline Mathlib/Tactic/CategoryTheory/Monoidal.lean:68]:
```suggestion
def mkContext? (e : Expr) : MetaM (Option Context) := do
```

I believe generally defs in `MetaM (Option ...)` should have a `?` at the end of their name.

File: Mathlib/Tactic/CategoryTheory/Monoidal.lean
Comment (joneugster) [inline Mathlib/Tactic/CategoryTheory/Monoidal.lean:562]: I don't yet understand what the implications of these changes are and why they are necessary. Could you quickly explain please? Do you know why there used to be a `catch` which isn't needed anymore?

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (joneugster) [inline Mathlib/Tactic/Widget/StringDiagram.lean:268]: Optional: Looking at the examples I had the idea that you could modify `fromExpr` with an optional argument so that in the case of equality, you get the text `no string diagram for RHS`/`.. for LHS`.

I was for a split-second confused why it displays one diagram plus "No string diagram" in one of the examples.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (joneugster) [inline Mathlib/Tactic/Widget/StringDiagram.lean:14]: I would appreciate if you expand the doc a little bit with what a string diagram is and what you exactly implemented.

Can be a very brief/rudimentary, maybe including an almost trivial example, or a reference. Just so that the unfamiliar reader has some starting point.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (joneugster) [inline Mathlib/Tactic/Widget/StringDiagram.lean:274]: Another optional suggestion, would it be simple to implement and useful to have a `=` between the two diagrams in the widget?  something like `<p> = </p>` with according style.

**Review by joneugster — COMMENTED** [review]
I will ask some reviewer more familiar with the widgets code to look at it again, too. But generally the PR seems to be in good shape and should be able to merge soon.

Left a few comments/questions.

File: Mathlib/Tactic/CategoryTheory/Monoidal.lean
Comment (yuma-mizuno) [reply Mathlib/Tactic/CategoryTheory/Monoidal.lean:562]: Previously I was using `ExprPresenter`, but then the normalization program had trouble generating proof terms for normalization. One reason was that the `MonoidalCategory C` instance in the assumptions was not available. In my understanding, tasks like generating proofs are beyond the scope of `ExprPresenter`.

In the current implementation of `StringDiagram.rpc` function, it explicitly deals with local instances using the `Meta.withLCtx` function.

File: test/StringDiagram.lean
Comment (yuma-mizuno) [reply test/StringDiagram.lean:21]: This is because there are no non-structural morphisms in the RHS. I've tried to improve the message to make the situation clearer.

Probably a better way would be to display the trivial vertical strings instead of the text massage, but this would require some extra work.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (yuma-mizuno) [reply Mathlib/Tactic/Widget/StringDiagram.lean:297]: It is possible to have a `=` between the diagrams, but it may not be an immediate task to do so in a good format. For example, we may want to use the same font size as the string diagram, but I'm not sure how to do this. To be honest, I don't know much about HTML or JavaScript.

Comment (yuma-mizuno) [reply Mathlib/Tactic/Widget/StringDiagram.lean:297]: I think adding the texts like "LHS" and "RHS" instead (as you suggested in another comment) would make it clearer that it is showing both sides of the equality.

File: Mathlib/Tactic/CategoryTheory/Monoidal.lean
Comment (Vtec234) [reply Mathlib/Tactic/CategoryTheory/Monoidal.lean:562]: > One reason was that the `MonoidalCategory C` instance in the assumptions was not available. In my understanding, tasks like generating proofs are beyond the scope of `ExprPresenter`.

Ah, no, that was an oversight in ProofWidgets. I pushed a fix for it now, and confirmed that your code also works with `@[expr_presenter]` using updated ProofWidgets.

The real difference between `ExprPresenter`s and the panel widget approach is that the latter is more flexible but needs more boilerplate. With an `ExprPresenter`, your function can only process a single expression when computing the display, as opposed to integrating data from multiple local hypotheses (for instance multiple assumed equalities) in a single display. For this price, you do not have to add extra handling for multiple goals or for visualizing hypotheses.

Since visualizing a single expression is what you are doing now, maybe the limitation is okay? Do you anticipate a need for the flexibility of panel widgets later?

More generally, the `ExprPresenter` API doesn't get used much and I'm not sure if it's even needed. It would be nice to see a real use so that it can be either improved or retired (if we decide it's not useful).

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (Vtec234) [inline Mathlib/Tactic/Widget/StringDiagram.lean:286]:
```suggestion
      g.ctx.val.runMetaM {} do
        g.mvarId.withContext do
```
Iirc these are equivalent.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (Vtec234) [inline Mathlib/Tactic/Widget/StringDiagram.lean:35]: This is a little confusing to me (but I am not an n-categorist)! Are objects not 0-morphisms, and morphisms 1-morphisms? Then an example of 2-morphisms would be the natural transformations in Cat.

Comment (yuma-mizuno) [reply Mathlib/Tactic/Widget/StringDiagram.lean]: Thank you for the comment. My explanation here wasn't very helpful. I added an explanation about the relationship between monoidal categories and bicategories and why I used bicategory terminology in this file.

File: Mathlib/Tactic/CategoryTheory/Monoidal.lean
Comment (yuma-mizuno) [reply Mathlib/Tactic/CategoryTheory/Monoidal.lean:562]: @Vtec234  Thanks for explanation! It's helpful. Also, thanks for the update.

I'm not sure if I need such flexibility in the string diagram at the moment. 

My impression of the `ExprPresenter` is that it is nice because it is enough intuitive to use for me.

**Review by joneugster — APPROVED** [review]
Thanks, the widgets look very good to me!

Maybe @Vtec234 can `maintainer-merge` if happy, too.

~~In the tests, what's the motivation for adding `SelectionPanel` now? The two widgets aren't interacting, are they? Adding it suggests to me that your widget would do something on shift-selecting subexpressions. (Unless I missed something, I'd suggest removing it again.)~~

**Comment by yuma-mizuno** [comment]
> In the tests, what's the motivation for adding `SelectionPanel` now? The two widgets aren't interacting, are they? Adding it suggests to me that your widget would do something on shift-selecting subexpressions. (Unless I missed something, I'd suggest removing it again.)

@joneugster  The `SelectionPanel` test is for verifying that the `ExprPresenter` works, but this only has meaning when the ProofWidget that mathlib depends on is updated to v0.0.40 (related comment: https://github.com/leanprover-community/mathlib4/pull/10581#discussion_r1688955775). In the current version (v0.0.39), `SelectionPanel` does not yields errors, but it does not display the string diagram.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (Vtec234) [inline Mathlib/Tactic/Widget/StringDiagram.lean:333]: Since `GoalTypePanel` can also draw this now using the presenter, should we get rid of this and the `StringDiagram` component?

**Review by Vtec234 — APPROVED** [review]
Looks awesome! One comment, and we will have to wait for #15100 to be merged first.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (Vtec234) [reply Mathlib/Tactic/Widget/StringDiagram.lean:340]: I see! Yes, there is no good way to debug `ExprPresenter`s right now. I am assuming that you need to keep the rpc method if you want to keep `StringDiagram`, as the rpc method is its main implementation!

**Comment by Vtec234** [comment]
I can't maintainer-merge (not a mathlib reviewer), but whoever can please do. LGTM!

**Comment by joneugster** [comment]
maintainer merge

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (mattrobball) [inline Mathlib/Tactic/Widget/StringDiagram.lean:67]: I wonder if there is any value to using something unboxed like `UInt` in place of `Nat`.

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (mattrobball) [inline Mathlib/Tactic/Widget/StringDiagram.lean:124]: Similarly, here and below, there is probably a performance benefit swapping out `List` for `Array` but I don't know what size the diagrams need to be before it is realized.

**Review by mattrobball — APPROVED** [review]
This is a great tool! I left some possibly useful comments that don't need to acted upon in this PR.

**Comment by mattrobball** [comment]
Since it has been a bit, @yuma-mizuno can you do a quick merge with master to make sure CI is happy before I send it to `bors`?

File: Mathlib/Tactic/Widget/StringDiagram.lean
Comment (Vtec234) [reply Mathlib/Tactic/Widget/StringDiagram.lean:124]: I would say no diagram you'd want to see in practice is large enough for this to make a meaningful difference. But I haven't benchmarked it, only tasseomancied.

Comment (mattrobball) [reply Mathlib/Tactic/Widget/StringDiagram.lean:124]: I agree. I think the direct `Expr` manipulation might be the best place to look for a real world improvement. This is also something that can be thought about afterwards.

**Comment by mattrobball** [comment]
bors merge

**Pull request closed without merging**
